### PR TITLE
Fix: #41250 You can't use a tablet while inside a locker

### LINKED ISSUE #41250: You can't use a tablet while inside a locker
Master

Hold a tablet, open the UI, step inside a locker, get locked out of the interface.

### PR DESCRIPTION
**Pull request by raffclar**
## About The Pull Request

Fixes #41250
Allows tablets to skip the interaction adjacency check for UI status updates if the tablet's location is the player/user.
Other computer forms such as laptops are still restricted for now.

## Why It's Good For The Game

This allows a tablet to be used inside a locker or other non-turf locations.

## Changelog

:cl:
fix: Tablets can be used inside lockers again.
/:cl:

### CODE REVIEW

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Gut says this should be universal, not completely sure about it tho, gut's fucked me over before

File: code/__DEFINES/interaction_flags.dm [hunk 1]
```diff
@@ -16,6 +16,8 @@
 #define INTERACT_ATOM_NO_FINGERPRINT_ATTACK_HAND (1<<7)
 /// adds hiddenprints instead of fingerprints on interact
 #define INTERACT_ATOM_NO_FINGERPRINT_INTERACT (1<<8)
+/// allows this atom to skip the adjacency check
+#define INTERACT_ATOM_ALLOW_IGNORE_ADJACENCY (1<<9)
```
[the remark this answers is not in the document]
Comment (raffclar) [reply code/__DEFINES/interaction_flags.dm:20]: Some new name ideas:
`INTERACT_ATOM_ALLOW_USER_LOCATION`
`INTERACT_ATOM_ALLOW_ADJACENT_OR_USER_LOCATION` 
Thoughts?

File: code/_onclick/other_mobs.dm [hunk 2]
```diff
@@ -83,7 +84,8 @@
 
 /atom/ui_status(mob/user)
 	. = ..()
-	if(!can_interact(user))
+	//Check if both user and atom are at the same location
+	if(!can_interact(user, !(user == src.loc)))
```
[the remark this answers is not in the document]
Comment (raffclar) [reply code/_onclick/other_mobs.dm:88]: I'm honestly terrified I'll blow something up so I limited the scope to just UI status checks. I could widen the gate and allow everything to be done if `user === src.loc` but we may have... unforseen consequences

File: code/_onclick/other_mobs.dm [hunk 3]
```diff
@@ -64,8 +64,9 @@
 		return interact(user)
 	return FALSE
 
-/atom/proc/can_interact(mob/user)
-	if(!user.can_interact_with(src))
+/atom/proc/can_interact(mob/user, require_adjacent_turf = TRUE)
+	var/requires_adjacency_check = !(interaction_flags_atom & INTERACT_ATOM_ALLOW_IGNORE_ADJACENCY)
+	if(!user.can_interact_with(src, requires_adjacency_check || require_adjacent_turf))
```
[the remark this answers is not in the document]
Comment (raffclar) [reply code/_onclick/other_mobs.dm:69]: It's a bit shit yeah.
Basically there's two conditions that need to be met for atom.loc == user.loc to be accepted as a stand-in for a full adjacency check.

1. Does the item have the `INTERACT_ATOM_ALLOW_IGNORE_ADJACENCY ` flag set?
2. Does the caller of can_interact allow `atom.loc == user.loc` to be accepted?

[the remark this answers is not in the document]
Comment (raffclar) [reply code/_onclick/other_mobs.dm:69]: @LemonInTheDark 
What about replacing `require_adjacent_turf = TRUE` with `allow_user_location = FALSE`?

**Comment by raffclar** [comment]
I have improved the wording, variables names. This interaction change is no longer limited to UI updates. It now encompasses all interactions. The interaction change is still specific to tablets/pdas. @LemonInTheDark Can you check my latest commit thanks

File: code/_onclick/other_mobs.dm [hunk 4]
```diff
@@ -64,8 +64,8 @@
 		return interact(user)
 	return FALSE
 
-/atom/proc/can_interact(mob/user)
-	if(!user.can_interact_with(src))
+/atom/proc/can_interact(mob/user, require_adjacent_turf = TRUE)
+	if(!user.can_interact_with(src, interaction_flags_atom & INTERACT_ATOM_ALLOW_USER_LOCATION))
```
Comment (LemonInTheDark) [inline code/_onclick/other_mobs.dm:68]: require_adjacent_turf doesn't need to exist anymore does it?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
sure.
re: apply to everything, I'm a bit worried even though it feels like the correct thing to do. Kinda unsure which is new

File: code/_onclick/other_mobs.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/_onclick/other_mobs.dm]: mood

File: code/_onclick/other_mobs.dm (hunk 4 above)
Comment (raffclar) [reply code/_onclick/other_mobs.dm:68]: Correct. Nice spot

**Comment by raffclar** [comment]
> sure. re: apply to everything, I'm a bit worried even though it feels like the correct thing to do. Kinda unsure which is new

Agreed. Perhaps a test merge?

**Comment by LemonInTheDark** [comment]
I'm gonna ask some other maints for opinions, see if I'm deranged or not. if yes absolutely a tm

**Comment by Mothblocks** [comment]
@LemonInTheDark 6 days, progress?

**Comment by LemonInTheDark** [comment]
Oh right yeah

**Comment by Mothblocks** [comment]
Lemon's on break so I'll just assume this is fine

**Comment by LemonInTheDark** [comment]
this was prob fine

**Pull request merged**